snakemake/persistence.py

```diff
@@ -434,9 +435,15 @@ def _output(self, job):
 
     def _record(self, subject, json_value, id):
         recpath = self._record_path(subject, id)
-        os.makedirs(os.path.dirname(recpath), exist_ok=True)
-        with open(recpath, "w") as f:
-            json.dump(json_value, f)
+        recdir = os.path.dirname(recpath)
+        os.makedirs(recdir, exist_ok=True)
+        # Write content to temporary file and rename it to the final file.
+        # This avoids race-conditions while writing (e.g. on NFS when the main job
+        # and the cluster node job propagate their content and the system has some
+        # latency including non-atomic propagation processes).
+        with tempfile.NamedTemporaryFile(mode="w", dir=recdir, delete=False) as tmpfile:
```

Comment:
What about adding `prefix=os.path.basename(recpath), suffix="."`? We would then get a similar behaviour as `rsync` has that the temporary files can be matched to the actual destinatio files.